Issue 250 complete permission checks

scottjones opened the pull request:
Fixes #250 
- Ensures you're an admin user to view/edit users
- Ensures you're an admin user to view/edit agencies
- Fixes editing users so that it works rather than just editing yourself
- Adds factory girl and request specs for validating permissions.

Note that I've added the factories all into a single file.  This is recommended by the folks from thoughtbot and is something that I'm finding I enjoy.

![image](https://cloud.githubusercontent.com/assets/444693/15276519/f2655316-1aaf-11e6-8901-b781a5a7dece.png)

![image](https://cloud.githubusercontent.com/assets/444693/15276525/17dba172-1ab0-11e6-9691-e7544ae1004b.png)

scottjones committed:
Add permission check on the admin/users page

* app/controllers/application_controller.rb: Add method for a
  permission error which we use in a requires_admin before filter.

* app/controllers/users_controller.rb: Require admin.

* app/views/errors/403.html.erb: Template for a permission denied
  error.

* spec/controllers/users_controller_spec.rb: Remove in favor of a new
  request spec.

* spec/rails_helper.rb: Register a new helper for mocking a sign_in.

* spec/request/users_spec.rb: New request spec to replace the
  controller spec for users controller.

* spec/support/request_spec_helper.rb: Adds the ability to sign_in as
  a user for a request spec.

scottjones committed:
Add another test for when the user isn't logged in

* spec/request/users_spec.rb: Ensure that when a user isn't logged in
  that they get redirected to sign in.

scottjones committed:
Make user editing work. Devise only allows editing yourself.

* app/controllers/users_controller.rb: Add edit action, and correct
  the parameters allowed to be updated.

* app/models/user.rb: Add a to_s method.

* app/views/users/index.html.erb: Link to the real "edit user" page.

* config/routes.rb: Add "edit user" route.

* spec/request/users_spec.rb Verify permissions on editing users.

scottjones committed:
Only admins can view agencies

* app/controllers/agencies_controller.rb: Only admins can view
  agencies.

* spec/request/agencies_spec.rb: Add tests to verify that only admins
  can view/edit agencies.

scottjones committed:
Whoops, forgot to commit edit view

* app/views/users/edit.html.erb: New view for editing users.

scottjones committed:
Ensure the grant being viewed/edited is accessible

* app/controllers/application_controller.rb: Add new not found error.

* app/controllers/budgets_controller.rb:
* app/controllers/forms_controller.rb:
* app/controllers/grants_controller.rb: Check that the grant is
  visible by the user when looking it up.  If not, don't find it and
  render the "not found"/404 error.

* app/views/errors/403.html.erb: Add button to go to grants list.

* app/views/errors/404.html.erb: New error page about not being able
  to find something.

scottjones committed:
Introduce factory girl, clean up users spec

* Gemfile
* Gemfile.lock: Add factory_girl

* spec/factories/users.rb: New factory for users.  Allows admin,
  unconfirmed and unapproved traits.

* spec/request/users_spec.rb: Use factories.

* spec/support/factory_girl.rb: Load factory girl syntax.

scottjones committed:
Use before_action instead of before_filter

scottjones committed:
Update agencies spec to use factory girl, add agency factory

* spec/factories/agencies.rb: New factory for agencies.

* spec/factories/users.rb: Add association for agencies from users.
  Admin user will be attached to HIF.  Add explicit trait for case
  workers to make tests clearer.

* spec/request/agencies_spec.rb: Update to use factories.

* spec/request/users_spec.rb: Update to use case worker trait to make
  test as clear as possible.

scottjones committed:
Expand factories and move to one file.  Add forms request specs.

* factories/users.rb:
* factories/agencies.rb:
* factories.rb: Combined individual factories into a single file, and
  added grants and other bits.

* models/agency_spec.rb:
* models/budget_spec.rb:
* models/grant_spec.rb:
* models/person_spec.rb
* request/forms_spec.rb: Use factory girl rather than creating objects
  directly as appropriate.

scottjones committed:
Merge branch 'master' into issue-250-complete-permission-checks

joedean commented on app/controllers/budgets_controller.rb:
```diff
@@ -52,6 +53,11 @@ def initialize_budgets_if_needed
+    @grant = Grant.visible_for_user(current_user).where(id: params[:grant_id]).first
+    not_found_error if @grant.nil?
+  end
+
```
would prefer this logic to be back in the model.

scottjones replied:
This seems like controller logic to me.  Find a grant that's visible with an id of X.  If not there, take them to the 404 page.  If `visible_for_user` was implemented here... that'd be a different story.

joedean commented on app/controllers/forms_controller.rb:
```diff
@@ -29,11 +29,13 @@ def update
+      else
+        Grant.visible_for_user(current_user).where(id: params[:grant_id]).first
+      end
+    not_found_error if @grant.nil?
```
model logic

joedean commented on app/controllers/budgets_controller.rb:
```diff
@@ -1,6 +1,7 @@
 class BudgetsController < ApplicationController
+  before_action :set_grant
```
I've tended to keep my controllers on the wetter side here to make code more readable and avoid additional before_actions in the controller.  This does mean a the same call to the model is required for all actions that require a grant to be set making code wet.  But the one call would all be in the action making the action easier to read, IMO.  This is my personal preference, what do you guys think?  At a min we should explicitly define before action only to apply to actions that require the grant to be set.  Doesn't make sense for "new" action, etc.

scottjones replied:
@joedean remember that this is the budget controller not the grant controller.  Every action for the budget needs the grant.

joedean commented on app/controllers/grants_controller.rb:
```diff
@@ -66,7 +66,8 @@ def destroy
-    @grant = Grant.find(params[:id])
+    @grant = Grant.visible_for_user(current_user).where(id: params[:id]).first
+    not_found_error if @grant.nil?
   end
```
model logic

joedean commented on Gemfile:
```diff
@@ -44,6 +44,7 @@ gem 'will_paginate-bootstrap'
 group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem "byebug"
+  gem 'factory_girl_rails'
```
awesome!  I did the same thing :-)  I also included the Faker gem

joedean commented on app/controllers/users_controller.rb:
```diff
@@ -1,14 +1,24 @@
 
+  def edit
+    @user = User.find params[:id]
+  end
```
Ah.  Awesome!  This is my preference to keep these instance variables being defined directly in the action without a before_action call.  :smile:

joedean commented on spec/factories.rb:
```diff
@@ -0,0 +1,66 @@
+    trait :unconfirmed do
+      confirmed_at nil
+    end
+  end
```
I have all my factories in different files under the factories directory.  I feel like that would be easier to maintain.  Would you agree?  I can convert these to files during my commit, if you would like.

scottjones replied:
I disagree.  Go ahead and watch the video by the thoughtbot guys.  https://thoughtbot.com/upcase/videos/factory-girl

Their argument is that if your factories are complex enough for them not to be in a single file then you're doing something with your factories that you shouldn't be...

joedean commented on spec/models/agency_spec.rb:
```diff
@@ -20,8 +12,8 @@
-      Agency.create(name: "Acme Agency")
+      create(:agency, name: "Zillow Agency")
+      create(:agency, name: "Acme Agency")
     end
```
You can use create_list instead of having to create one agency at a time.  E.g., FactoryGirl.create_list :agency, 3, name: { Faker::Company.name }

joedean commented on spec/models/grant_spec.rb:
```diff
@@ -2,16 +2,13 @@
+    let(:grant1) { create(:grant, user: worker1) }
+    let(:grant2) { create(:grant, user: worker2) }
+    let(:grant3) { create(:grant, user: worker3) }
     before do
```
I created a factory called agency_with_grants, number_of grants.  This could be used hear to avoid the setup in the test.  I can also refactor when my stuff gets put in.

joedean commented on spec/models/grant_spec.rb:
```diff
@@ -74,7 +72,7 @@
+        grants = Grant.list(worker1, agency_id: grant3.user.agency.id)
         expect(grants).to_not include(grant3)
       end
     end
```
FYI - I did some refactoring on this test.  Added some more unit tests.  I can integrate when my stuff goes in.

joedean commented on spec/models/person_spec.rb:
```diff
@@ -1,12 +1,12 @@
+      person.incomes = [build(:income, :previous, monthly_income: 1000),
+                        build(:income, monthly_income: 2000),
+                        build(:income, :disability, monthly_income: 3000)]
     end
```
you can also use build_list

scottjones replied:
Doesn't that build 3 of the same exact kind of object?   Here we want to vary whether they're `previous`, `disability` and the `amount`s...

joedean commented on spec/support/factory_girl.rb:
```diff
@@ -0,0 +1,4 @@
+# Include factory girl
+RSpec.configure do |config|
+  config.include FactoryGirl::Syntax::Methods
+end
```
Great!  I didn't do this and was just referencing FactoryGirl.  Thanks!

joedean commented:
I added comments in the code.  Overall looks great!  Thanks for adding all the request specs and adding FactoryGirl was a must!  :smiley:

joedean commented:
@lamplus I'm leaving you in charge of merging in all pull requests, since you are the maintainer of the repo.

scottjones committed:
Add custom errors and use them.  Move find logic for grants to model

* app/controllers/application_controller.rb: Rescue errors and render
  the appropriate views.

* app/controllers/budgets_controller.rb:
* app/controllers/forms_controller.rb:
* app/controllers/grants_controller.rb: Use new Grant.find_if_visible
  method to look up grant.

* app/models/errors/not_authorized_error.rb:
* app/models/errors/not_found_error.rb: Custom error exceptions.

* app/models/grant.rb: Add new find_if_visible method.  Raise
  exception if we couldn't find it.

* app/views/errors/404.html.erb -> app/views/errors/not_found.html.erb:
* app/views/errors/403.html.erb -> app/views/errors/unauthorized.html.erb:
  Rename

* spec/request/agencies_spec.rb:
* spec/request/forms_spec.rb:
* spec/request/users_spec.rb: Assert with named statuses rather than
  numbers (note that I had 403 rather than 401!).

scottjones committed:
Don't use static error pages

* app/controllers/errors_controller.rb: New controller for the various
  error pages.

* app/views/errors/internal_server_error.html.erb: New page for when
  errors occur.

* config/application.rb:
* config/routes.rb: Specify that we're using our own routes for
  errors.

* public/404.html:
* public/500.html: Remove static error pages.

scottjones commented:
@lamplus pushed a few more commits here that should address @joedean's comments.

lamplus committed:
Merge branch 'master' into issue-250-complete-permission-checks
